### PR TITLE
feat: leave chat #301

### PR DESCRIPTION
**Pull request by crui3er**
Leave chat button is accessiable from author settings modal.
For the moment only not owners can leave a chat.
image
image

### CODE REVIEW

**Review by alexis-kochetov — APPROVED** [review]

File: src/dotnet/Chat.UI.Blazor/Components/AuthorSettingsModal.razor [hunk 1]
```diff
@@ -8,6 +8,9 @@
             <UserNameEditor/>
         </section>
 
+        <h4 class="section"></h4>
```
Comment (FrolYo) [inline src/dotnet/Chat.UI.Blazor/Components/AuthorSettingsModal.razor:11]: why empty? maybe remove?

File: src/dotnet/Chat.Service/ChatAuthorsBackend.cs [hunk 2]
```diff
@@ -189,6 +189,45 @@ public virtual async Task<ChatAuthor> Create(IChatAuthorsBackend.CreateCommand c
...
+            _ = ListAuthorIds(chatId, default);
+
+            return default!;
+        }
+
+        var dbContext = await CreateCommandDbContext(cancellationToken).ConfigureAwait(false);
+        await using var __ = dbContext.ConfigureAwait(false);
+
+        var dbChatAuthor = await dbContext.ChatAuthors
+            .SingleOrDefaultAsync(a => a.Id == authorId, cancellationToken)
```
Comment (FrolYo) [inline src/dotnet/Chat.Service/ChatAuthorsBackend.cs:217]: do we really need `SingleOrDefaultAsync` here?
https://docs.microsoft.com/en-us/aspnet/core/data/ef-rp/crud?view=aspnetcore-6.0#ways-to-read-one-entity

File: src/dotnet/Chat.Service/Chats.cs [hunk 3]
```diff
@@ -412,7 +441,24 @@ private async Task<ChatEntry> GetChatEntry(Session session, string chatId, long
...
     private async Task JoinChat(Session session, string chatId, CancellationToken cancellationToken)
-        => await ChatAuthorsBackend.GetOrCreate(session, chatId, cancellationToken).ConfigureAwait(false);
+    {
+        var chatAuthor = await ChatAuthorsBackend.GetOrCreate(session, chatId, cancellationToken).ConfigureAwait(false);
+        if (chatAuthor.HasLeft) {
+            var command = new IChatAuthorsBackend.ChangeHasLeftCommand(chatAuthor.Id, false);
+            await Commander.Call(command, cancellationToken).ConfigureAwait(false);
+        }
+        if (await IsInvited(session, chatId, cancellationToken).ConfigureAwait(false)) {
+            var updateOptionCommand1 = new ISessionOptionsBackend.UpsertCommand(
```
Comment (FrolYo) [inline src/dotnet/Chat.Service/Chats.cs:451]: why not pass multiple options into a single command? I see that https://github.com/Actual-Chat/actual-chat/blob/3381910bcc566e716dd51e8fb8f6b038f6242acc/src/dotnet/Users.Service/SessionOptionsBackend.cs#L28 can support multiple options - we just need to pass them into command instance

**Review by FrolYo — CHANGES_REQUESTED** [review]

**Review by FrolYo — CHANGES_REQUESTED** [review]

File: src/dotnet/Chat.UI.Blazor/Components/AuthorSettingsModal.razor (hunk 1 above)
Comment (crui3er) [reply src/dotnet/Chat.UI.Blazor/Components/AuthorSettingsModal.razor:11]: it gives separation line

File: src/dotnet/Chat.Service/Chats.cs [hunk 4]
```diff
@@ -352,6 +354,33 @@ public virtual async Task RemoveTextEntry(IChats.RemoveTextEntryCommand command,
...
+            return;
+        var chat = await Get(session, chatId, cancellationToken).ConfigureAwait(false);
+        if (chat == null)
+            return;
+
+        var user = await Auth.GetUser(session, cancellationToken).ConfigureAwait(false);
+        if (chat.OwnerIds.Contains(user.Id)) {
+            throw new InvalidOperationException(
+                    "You are the last owner of the chat. At the moment delete and leave feature is missing.");
+            // TODO: managing ownership functionality is required
```
[the remark this answers is not in the document]
Comment (crui3er) [reply src/dotnet/Chat.Service/Chats.cs:374]: Better to do it separetly

File: src/dotnet/Chat.Contracts/Backend/IChatAuthorsBackend.cs [hunk 5]
```diff
@@ -23,9 +23,18 @@ public interface IChatAuthorsBackend : IComputeService
...
+    Task<ChatAuthor> ChangeHasLeft(ChangeHasLeftCommand command, CancellationToken cancellationToken);
+
     [DataContract]
     public sealed record CreateCommand(
         [property: DataMember] string ChatId,
         [property: DataMember] string UserId
         ) : ICommand<ChatAuthor>, IBackendCommand;
+
+    [DataContract]
+    public sealed record ChangeHasLeftCommand(
```
Comment (alexyakunin) [inline src/dotnet/Chat.Contracts/Backend/IChatAuthorsBackend.cs:36]: Ideally there should be just one UpdateCommand, which applies all proposed changes. Proposed change = field value of command != null, which describes it.

The reasoning is simple: most of these actions will share nearly the same logic for pulling entity, running checks, etc.; same for the front-end. So why bother to create N commands if you can have just one?

Besides that, for any complex change you can create fields like `MoveAction? Move { get; init; }` inside `UpdateCommand`.

**Review by alexyakunin — CHANGES_REQUESTED** [review]

File: src/dotnet/Chat.Contracts/Backend/IChatAuthorsBackend.cs (hunk 5 above)
Comment (alexyakunin) [reply src/dotnet/Chat.Contracts/Backend/IChatAuthorsBackend.cs:36]: Long story short, CRUD, where U can be responsible for all possible updates.

File: src/dotnet/Chat.Service/Chats.cs [hunk 6]
```diff
@@ -352,6 +354,33 @@ public virtual async Task RemoveTextEntry(IChats.RemoveTextEntryCommand command,
...
+        if (chatAuthor == null || chatAuthor.HasLeft)
+            return;
+        var chat = await Get(session, chatId, cancellationToken).ConfigureAwait(false);
+        if (chat == null)
+            return;
+
+        var user = await Auth.GetUser(session, cancellationToken).ConfigureAwait(false);
+        if (chat.OwnerIds.Contains(user.Id)) {
+            throw new InvalidOperationException(
+                    "You are the last owner of the chat. At the moment delete and leave feature is missing.");
```
Comment (alexyakunin) [inline src/dotnet/Chat.Service/Chats.cs:373]: InvalidOperationException -> NotSupportedException, + text = "The very last owner of the chat can't leave it."

File: src/dotnet/Chat.UI.Blazor/Components/LeaveChatConfirmationModal.razor [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            <Button Class="btn-cancel-outline" Click="OnCancel">Cancel</Button>
+            <Button Class="btn-error" Click="OnLeave">Leave</Button>
+        </div>
+    </Buttons>
+</DialogFrame>
+
+@code {
+    [Inject] Session Session { get; init; } = null!;
+    [Inject] UICommandRunner Commander { get; init; } = null!;
+    [CascadingParameter] BlazoredModalInstance ModalInstance { get; set; } = null!;
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/LeaveChatConfirmationModal.razor:20]: Let's declare all parameters as public. Not sure why this works, but I recently run S&R to add "public" to all of such properties.

File: src/dotnet/Db/Module/DbModule.cs [hunk 8]
```diff
@@ -58,8 +61,61 @@ public class DbModule : HostModule<DbSettings>
         };
 
         // Adding services
+        services.TryAddSingleton<OutstandingDbConnectionsCounter>();
         services.AddSingleton(dbInfo);
-        services.AddDbContextFactory<TDbContext>(builder => {
+        services.AddDbContextFactory<TDbContext>((svp, builder) => {
+            var counter = svp.GetRequiredService<OutstandingDbConnectionsCounter>();
```
Comment (alexyakunin) [inline src/dotnet/Db/Module/DbModule.cs:67]: I would actually move this code to some dedicated type.

File: src/dotnet/Db/Module/DbModule.cs [hunk 9]
```diff
@@ -109,6 +165,51 @@ public class DbModule : HostModule<DbSettings>
         });
     }
 
+    private static string DumpConnections(DbConnection dbConnection)
```
Comment (alexyakunin) [inline src/dotnet/Db/Module/DbModule.cs:168]: Same for this piece.

Comment (alexyakunin) [reply src/dotnet/Db/Module/DbModule.cs:168]: DbModule is definitely not the right place for this.

**Review by alexyakunin — CHANGES_REQUESTED** [review]

File: src/dotnet/Db/Module/DbModule.cs (hunk 8 above)
Comment (crui3er) [reply src/dotnet/Db/Module/DbModule.cs:67]: I will remove all this code from DbModule after I complete connections leak investigation.

File: src/dotnet/Db/Module/DbModule.cs (hunk 9 above)
Comment (crui3er) [reply src/dotnet/Db/Module/DbModule.cs:168]: this code is reverted

File: src/dotnet/Chat.Service/Chats.cs (hunk 6 above)
Comment (crui3er) [reply src/dotnet/Chat.Service/Chats.cs:373]: fixed

File: src/dotnet/Chat.UI.Blazor/Components/AuthorSettingsModal.razor [hunk 10]
```diff
@@ -18,8 +21,14 @@
 
 @code {
     [Inject] private ComponentIdGenerator ComponentIdGenerator { get; init; } = null!;
+    [Inject] private ModalUI ModalUI { get; init; } = null!;
     [CascadingParameter] BlazoredModalInstance ModalInstance { get; set; } = null!;
     [Parameter] public Model ModalModel { get; set; } = null!;
 
     public sealed record Model(Chat Chat);
+
+    private void OnLeaveChatClicked() {
```
[the remark this answers is not in the document]
Comment (crui3er) [reply src/dotnet/Chat.UI.Blazor/Components/AuthorSettingsModal.razor:30]: fixed

File: src/dotnet/Chat.Contracts/Backend/IChatAuthorsBackend.cs (hunk 5 above)
Comment (crui3er) [reply src/dotnet/Chat.Contracts/Backend/IChatAuthorsBackend.cs:36]: Would you propose to define command like below and adapt handler?
```
[DataContract]
public sealed record UpdateCommand(
     [property: DataMember] string AuthorId,
     [property: DataMember] bool? HasLeft
) : ICommand<ChatAuthor>, IBackendCommand;
```

**Pull request merged**